executor/index_lookup_hash_join.go

```diff
@@ -0,0 +1,348 @@
+			break
+		}
+		err := iw.handleTask(ctx, task, joinResult)
+		if err != nil {
+			joinResult.err = err
+			break
+		}
+	}
+	if joinResult.err != nil || (joinResult.chk != nil && joinResult.chk.NumRows() > 0) {
+		iw.joinResultCh <- joinResult
+	}
+}
+
+func (iw *innerHashWorker) getNewJoinResult() (bool, *indexLookUpResult) {
+	joinResult := &indexLookUpResult{
+		src: iw.joinChkResourceCh[iw.workerID],
+	}
+	ok := true
+	select {
+	case joinResult.chk, ok = <-iw.joinChkResourceCh[iw.workerID]:
+	}
+	return ok, joinResult
+}
+func (iw *innerHashWorker) handleTask(ctx context.Context, task *lookUpJoinTask, joinResult *indexLookUpResult) error {
+	dLookUpKeys, err := iw.constructDatumLookupKeys(task)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = iw.fetchInnerResults(ctx, task, dLookUpKeys)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	var ok bool
+	ok, joinResult = iw.join2Chunk(joinResult, task)
+	if !ok {
+		return errors.New("join2Chunk failed")
+	}
+	it := task.lookupMap.NewIterator()
+	for key, rowPtr := it.Next(); key != nil; key, rowPtr = it.Next() {
+		iw.matchPtrBytes = task.matchKeyMap.Get(key, iw.matchPtrBytes[:0])
+		if len(iw.matchPtrBytes) == 0 {
+			ptr := *(*uint32)(unsafe.Pointer(&rowPtr[0]))
+			misMatchedRow := task.outerResult.GetRow(int(ptr))
+			iw.joiner.onMissMatch(misMatchedRow, joinResult.chk)
+		}
+		if joinResult.chk.NumRows() == iw.maxChunkSize {
+			ok := true
+			iw.joinResultCh <- joinResult
+			ok, joinResult = iw.getNewJoinResult()
+			if !ok {
+				return nil
+			}
+		}
+	}
+
+	return nil
+
+}
+func (iw *innerHashWorker) join2Chunk(joinResult *indexLookUpResult, task *lookUpJoinTask) (ok bool, _ *indexLookUpResult) {
+
+	for i := 0; i < task.innerResult.NumChunks(); i++ {
+		curChk := task.innerResult.GetChunk(i)
+		iter := chunk.NewIterator4Chunk(curChk)
+		for innerRow := iter.Begin(); innerRow != iter.End(); innerRow = iter.Next() {
+			ok, joinResult = iw.joinMatchInnerRow2Chunk(innerRow, task, joinResult)
+			if !ok {
+				return false, joinResult
+			}
+
+		}
+	}
+	return true, joinResult
+}
+
+func (iw *innerHashWorker) joinMatchInnerRow2Chunk(innerRow chunk.Row, task *lookUpJoinTask,
+	joinResult *indexLookUpResult) (bool, *indexLookUpResult) {
+	keyBuf := make([]byte, 0, 64)
+	for _, keyCol := range iw.keyCols {
+		d := innerRow.GetDatum(keyCol, iw.rowTypes[keyCol])
+		var err error
+		keyBuf, err = codec.EncodeKey(iw.ctx.GetSessionVars().StmtCtx, keyBuf, d)
+		if err != nil {
+			return false, joinResult
+		}
+	}
+	iw.matchPtrBytes = task.lookupMap.Get(keyBuf, iw.matchPtrBytes[:0])
+	if len(iw.matchPtrBytes) == 0 {
+		return true, joinResult
+	}
+	task.matchedInners = task.matchedInners[:0]
+	var matchedOuters []chunk.Row
+	for _, b := range iw.matchPtrBytes {
+		ptr := *(*uint32)(unsafe.Pointer(&b[0]))
+		matchedOuter := task.outerResult.GetRow(int(ptr))
+		matchedOuters = append(matchedOuters, matchedOuter)
+	}
+	innerIter := chunk.NewIterator4Slice([]chunk.Row{innerRow})
```

Comment:
It executes very slowly here...The `tryToMatch` function only match just one row. Can we change it by matching `outerRow`s to `innerRow`?